File: java/src/test/java/com/nexa/protocol/master/netty/handler/SessionIdentityTest.java
package com.nexa.protocol.master.netty.handler;

import com.nexa.protocol.master.EnvelopeFixtures;
import com.nexa.protocol.master.RecordingMasterListener;
import com.nexa.protocol.master.RunnerSession;
import com.nexa.protocol.master.SessionManager;
import com.nexa.protocol.master.netty.MasterChannelHandler;
import io.netty.channel.embedded.EmbeddedChannel;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * 协议侧身份校验测试（专项修复 A）：
 * 所有已注册消息必须以「发送 channel 绑定的会话」为准，报文 runnerId / source_id 仅作一致性检查。
 */
class SessionIdentityTest {

    private static final String TOKEN = "good-token";

    private SessionManager sessionManager;
    private RecordingMasterListener listener;

    @BeforeEach
    void setUp() {
        sessionManager = new SessionManager();
        listener = new RecordingMasterListener(TOKEN);
    }

    // ---- 心跳 ----

    /** 未注册连接冒用已在线节点身份发心跳：不刷新任何会话、不回包、不回调 */
    @Test
    void unregisteredConnectionCannotRefreshHeartbeat() {
        EmbeddedChannel online = newMasterChannel();
        register(online, "B");
        RunnerSession sessionB = sessionManager.get("B").orElseThrow();
        long before = sessionB.getLastHeartbeatTime();

        EmbeddedChannel rogue = newMasterChannel();
        rogue.writeInbound(EnvelopeFixtures.heartbeat("B").toByteArray());
        rogue.runPendingTasks();

        assertEquals(before, sessionB.getLastHeartbeatTime(),
                "unregistered heartbeat must not refresh session B");
        assertTrue(listener.heartbeats.isEmpty(), "unregistered heartbeat must not reach business listener");
        assertNull(rogue.readOutbound(), "unregistered connection must not receive a heartbeat response");
        assertSame(sessionB, sessionManager.get("B").orElseThrow());
    }

    /** 已注册 A 冒用 B 的 runnerId 发心跳：B 与 A 的心跳时间都不变，且不回包 */
    @Test
    void registeredConnectionCannotHeartbeatAsAnotherRunner() {
        EmbeddedChannel channelA = newMasterChannel();
        register(channelA, "A");
        EmbeddedChannel channelB = newMasterChannel();
        register(channelB, "B");

        RunnerSession sessionA = sessionManager.get("A").orElseThrow();
        RunnerSession sessionB = sessionManager.get("B").orElseThrow();
        long beforeA = sessionA.getLastHeartbeatTime();
        long beforeB = sessionB.getLastHeartbeatTime();

        channelA.writeInbound(EnvelopeFixtures.heartbeat("B").toByteArray());
        channelA.runPendingTasks();

        assertEquals(beforeB, sessionB.getLastHeartbeatTime(), "forged heartbeat must not refresh B");
        assertEquals(beforeA, sessionA.getLastHeartbeatTime(), "forged heartbeat must not refresh A either");
        assertTrue(listener.heartbeats.isEmpty(), "forged heartbeat must not reach business listener");
        assertNull(channelA.readOutbound(), "forged heartbeat must not be answered");
        assertNull(channelB.readOutbound(), "B must not receive a heartbeat response for A's forgery");
    }

    /** 正常心跳：刷新自身会话并只回发给发送方 */
    @Test
    void normalHeartbeatRefreshesOwnSessionAndRepliesToSender() {
        EmbeddedChannel channel = newMasterChannel();
        register(channel, "B");
        RunnerSession sessionB = sessionManager.get("B").orElseThrow();

        channel.writeInbound(EnvelopeFixtures.heartbeat("B").toByteArray());
        channel.runPendingTasks();

        assertEquals(List.of("B"), listener.heartbeats);
        assertNotNull(channel.readOutbound(), "sender should receive the heartbeat response");
        assertFalse(sessionB.isTimedOut());
    }

    // ---- 主动断开 ----

    /** 未注册连接冒用 B 身份发 DISCONNECT：B 会话不受影响，业务层不收到 B 的断开 */
    @Test
    void unregisteredConnectionCannotDisconnectRunner() {
        EmbeddedChannel online = newMasterChannel();
        register(online, "B");
        RunnerSession sessionB = sessionManager.get("B").orElseThrow();
        listener.reset();

        EmbeddedChannel rogue = newMasterChannel();
        rogue.writeInbound(EnvelopeFixtures.disconnect("B", "bye").toByteArray());
        rogue.runPendingTasks();

        assertSame(sessionB, sessionManager.get("B").orElseThrow(), "B session must survive forged disconnect");
        assertTrue(sessionB.isActive());
        assertEquals(0, listener.countDisconnects("B"), "business listener must not see B disconnect");
        assertFalse(rogue.isOpen(), "unregistered connection should be closed");
    }

    /** 已注册 A 冒用 B 身份发 DISCONNECT：只影响 A 自己，绝不移除 B 的会话 */
    @Test
    void registeredConnectionCannotDisconnectAnotherRunner() {
        EmbeddedChannel channelA = newMasterChannel();
        register(channelA, "A");
        EmbeddedChannel channelB = newMasterChannel();
        register(channelB, "B");
        RunnerSession sessionB = sessionManager.get("B").orElseThrow();
        listener.reset();

        channelA.writeInbound(EnvelopeFixtures.disconnect("B", "bye").toByteArray());
        channelA.runPendingTasks();

        assertSame(sessionB, sessionManager.get("B").orElseThrow(), "forged disconnect must not remove B");
        assertTrue(sessionB.isActive());
        assertEquals(0, listener.countDisconnects("B"));
        // A 自己的连接被拒绝关闭，其会话按正常断线清理
        assertFalse(channelA.isOpen());
        assertTrue(sessionManager.get("A").isEmpty(), "A's own session is cleaned up on its close");
    }

    /** 正常主动断开：条件移除自身会话并通知一次 */
    @Test
    void normalDisconnectRemovesOwnSessionAndNotifiesOnce() {
        EmbeddedChannel channel = newMasterChannel();
        register(channel, "A");
        listener.reset();

        channel.writeInbound(EnvelopeFixtures.disconnect("A", "shutdown").toByteArray());
        channel.runPendingTasks();

        assertTrue(sessionManager.get("A").isEmpty(), "session should be removed on normal disconnect");
        assertEquals(1, listener.countDisconnects("A"));
        assertEquals(List.of("shutdown"), listener.disconnectReasons("A"));
        assertFalse(channel.isOpen());
    }

    // ---- 回执类 ----

    /** 已注册 A 冒用 B 身份发任务回执：业务层不以 B 身份处理 */
    @Test
    void registeredConnectionCannotForgeTaskResult() {
        EmbeddedChannel channelA = newMasterChannel();
        register(channelA, "A");
        EmbeddedChannel channelB = newMasterChannel();
        register(channelB, "B");
        listener.reset();

        channelA.writeInbound(EnvelopeFixtures.taskResult("B", "task-1").toByteArray());
        channelA.runPendingTasks();

        assertTrue(listener.taskResults.isEmpty(), "forged task result must not be processed as B");
    }

    /** 已注册 A 冒用 B 身份发状态/日志回执：业务层不以 B 身份处理 */
    @Test
    void registeredConnectionCannotForgeContainerQueryResults() {
        EmbeddedChannel channelA = newMasterChannel();
        register(channelA, "A");
        EmbeddedChannel channelB = newMasterChannel();
        register(channelB, "B");
        listener.reset();

        channelA.writeInbound(EnvelopeFixtures.containerStatus("B", true, "running").toByteArray());
        channelA.writeInbound(EnvelopeFixtures.containerLogs("B", "logs").toByteArray());
        channelA.runPendingTasks();

        assertTrue(listener.containerStatuses.isEmpty(), "forged status must not be processed as B");
        assertTrue(listener.containerLogs.isEmpty(), "forged logs must not be processed as B");
    }

    /** 已注册 A 用 B 的 source_id 请求产物：按 channel 绑定识别为 A，不以 B 身份回调 */
    @Test
    void registeredConnectionCannotForgeArtifactRequestIdentity() {
        EmbeddedChannel channelA = newMasterChannel();
        register(channelA, "A");
        EmbeddedChannel channelB = newMasterChannel();
        register(channelB, "B");
        listener.reset();

        // source_id 声明 B，但发送方是 A 的连接
        channelA.writeInbound(EnvelopeFixtures.artifactRequest("B", "svc-1").toByteArray());
        channelA.runPendingTasks();

        assertTrue(listener.artifactRequests.isEmpty(),
                "artifact request with mismatched source_id must be rejected");
    }

    /** 正常产物请求：按发送 channel 绑定识别身份 */
    @Test
    void normalArtifactRequestIsAcceptedForOwningSession() {
        EmbeddedChannel channel = newMasterChannel();
        register(channel, "A");
        listener.reset();

        channel.writeInbound(EnvelopeFixtures.artifactRequest("A", "svc-1").toByteArray());
        channel.runPendingTasks();

        assertEquals(List.of("A"), listener.artifactRequests);
    }

    // ---- 接管后的旧连接 ----

    /** 旧连接被同 ID 新连接接管后继续发送已注册消息：一律拒绝，新会话不受影响 */
    @Test
    void supersededConnectionMessagesAreRejected() {
        EmbeddedChannel oldChannel = newMasterChannel();
        register(oldChannel, "R");

        EmbeddedChannel newChannel = newMasterChannel();
        // 手工构造「新连接已接管、旧连接尚未关闭」的交错状态
        RunnerSession superseding = new RunnerSession("R", newChannel, "host", "127.0.0.1", "test");
        sessionManager.register(superseding);
        SessionResolver.bind(newChannel, "R");
        listener.reset();

        assertTrue(oldChannel.isOpen(), "precondition: old connection still open but superseded");
        assertSame(superseding, sessionManager.get("R").orElseThrow(),
                "precondition: registry already points to the superseding session");

        long before = superseding.getLastHeartbeatTime();

        oldChannel.writeInbound(EnvelopeFixtures.heartbeat("R").toByteArray());
        oldChannel.writeInbound(EnvelopeFixtures.taskResult("R", "task-1").toByteArray());
        oldChannel.writeInbound(EnvelopeFixtures.disconnect("R", "bye").toByteArray());
        oldChannel.runPendingTasks();

        assertTrue(listener.heartbeats.isEmpty(), "superseded connection heartbeat must be rejected");
        assertTrue(listener.taskResults.isEmpty(), "superseded connection result must be rejected");
        assertEquals(0, listener.countDisconnects("R"), "superseded connection must not disconnect R");
        assertEquals(before, superseding.getLastHeartbeatTime(), "new session must not be refreshed by old connection");
        assertSame(superseding, sessionManager.get("R").orElseThrow(), "new session must stay registered");
        assertNull(newChannel.readOutbound(), "no response should be sent to the superseded connection");
    }

    /** 接管后新会话一切正常 */
    @Test
    void supersedingSessionKeepsWorking() {
        EmbeddedChannel oldChannel = newMasterChannel();
        register(oldChannel, "R");
        EmbeddedChannel newChannel = newMasterChannel();
        register(newChannel, "R");

        assertFalse(oldChannel.isOpen(), "old connection should be closed by takeover");
        RunnerSession current = sessionManager.get("R").orElseThrow();
        assertSame(newChannel, current.getChannel());

        listener.reset();
        newChannel.writeInbound(EnvelopeFixtures.heartbeat("R").toByteArray());
        newChannel.runPendingTasks();

        assertEquals(List.of("R"), listener.heartbeats, "new session should keep heartbeating");
        assertNotNull(newChannel.readOutbound());
    }

    /** 旧连接关闭不得误报新会话离线 */
    @Test
    void oldConnectionCloseDoesNotReportNewSessionOffline() {
        EmbeddedChannel oldChannel = newMasterChannel();
        register(oldChannel, "R");
        EmbeddedChannel newChannel = newMasterChannel();
        register(newChannel, "R");
        listener.reset();

        assertEquals(0, listener.countDisconnects("R"),
                "closing the superseded connection must not report R offline");
        assertTrue(sessionManager.get("R").orElseThrow().isActive());
    }

    // ---- helpers ----

    private EmbeddedChannel newMasterChannel() {
        MessageDispatcher dispatcher = new MessageDispatcher(List.of(
                new RegisterHandler(sessionManager, listener),
                new HeartbeatHandler(sessionManager, listener),
                new DisconnectHandler(sessionManager, listener),
                new TaskResultHandler(sessionManager, listener),
                new ContainerStatusHandler(sessionManager, listener),
                new ContainerLogsHandler(sessionManager, listener),
                new ArtifactHandler(sessionManager, listener)));
        MasterChannelHandler masterHandler = new MasterChannelHandler(sessionManager, listener, dispatcher);
        return new EmbeddedChannel(masterHandler);
    }

    private void register(EmbeddedChannel channel, String runnerId) {
        channel.writeInbound(EnvelopeFixtures.register(runnerId, TOKEN).toByteArray());
        channel.runPendingTasks();
        // 消费注册响应，便于后续对 outbound 的断言
        Object registerResponse = channel.readOutbound();
        assertNotNull(registerResponse, "register response expected for " + runnerId);
        assertTrue(sessionManager.get(runnerId).isPresent(), "session should be registered: " + runnerId);
    }
}

File: go/master/session_identity_test.go
package master

import (
	"log/slog"
	"testing"
	"time"

	"github.com/leehaohaohao/nexa-protocol/go/codec"
	"github.com/leehaohaohao/nexa-protocol/go/messages"
)

const goodToken = "good-token"

func testLogger() *slog.Logger {
	return slog.New(slog.DiscardHandler)
}

// newDispatchedMaster 组装与 NexaMaster 相同的 handler 集合
func newDispatchedMaster(sessions *SessionManager, listener Listener) *MessageDispatcher {
	return NewMessageDispatcher([]Handler{
		NewRegisterHandler(sessions, listener, testLogger()),
		NewHeartbeatHandler(sessions, listener, testLogger()),
		NewDisconnectHandler(sessions, listener, testLogger()),
		NewContainerStatusHandler(sessions, listener, testLogger()),
		NewContainerLogsHandler(sessions, listener, testLogger()),
		NewArtifactHandler(sessions, listener, testLogger()),
	}, testLogger())
}

// registerConn 通过 RegisterHandler 完成一次注册，返回该连接的 ConnContext
func registerConn(t *testing.T, dispatcher *MessageDispatcher, sessions *SessionManager, conn *mockConn, runnerId string) *ConnContext {
	t.Helper()

	connCtx := &ConnContext{Session: NewRunnerSession("", conn, "", "", "")}
	dispatcher.Dispatch(connCtx, registerEnvelope(runnerId, goodToken))

	if _, ok := sessions.Get(runnerId); !ok {
		t.Fatalf("session %q should be registered", runnerId)
	}
	return connCtx
}

func registerEnvelope(runnerId, token string) *messages.Envelope {
	return codec.BuildRegisterRequestWithToken(runnerId, runnerId+"-host", "127.0.0.1", "test", token)
}

func heartbeatEnvelope(runnerId string) *messages.Envelope {
	return codec.BuildHeartbeatRequest(runnerId, 0, 0, 0)
}

func disconnectEnvelope(runnerId, reason string) *messages.Envelope {
	return codec.BuildDisconnectRequest(runnerId, reason)
}

func statusEnvelope(runnerId string) *messages.Envelope {
	resp := &messages.ContainerStatusResponse{RunnerId: runnerId, Running: true, Status: "running"}
	return codec.BuildContainerStatusResponse("req-1", runnerId, resp)
}

func logsEnvelope(runnerId string) *messages.Envelope {
	resp := &messages.ContainerLogsResponse{RunnerId: runnerId, Content: "logs"}
	return codec.BuildContainerLogsResponse("req-1", runnerId, resp)
}

func artifactEnvelope(sourceId, serviceId string) *messages.Envelope {
	req := &messages.ArtifactRequest{ServiceId: serviceId, Type: "JAR", Version: 0}
	return codec.BuildArtifactRequest(sourceId, req)
}

// ---- 心跳身份校验 ----

// 未注册连接冒用已在线节点身份发心跳：不刷新任何会话、不回包、不回调
func TestUnregisteredConnectionCannotRefreshHeartbeat(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	onlineConn := newMockConn()
	registerConn(t, dispatcher, sessions, onlineConn, "B")
	sessionB, _ := sessions.Get("B")
	before := sessionB.lastHeartbeatTime.Load()

	rogueConn := newMockConn()
	rogueCtx := &ConnContext{Session: NewRunnerSession("", rogueConn, "", "", "")}
	dispatcher.Dispatch(rogueCtx, heartbeatEnvelope("B"))

	if got := sessionB.lastHeartbeatTime.Load(); got != before {
		t.Fatalf("unregistered heartbeat must not refresh session B: before=%d after=%d", before, got)
	}
	if listener.heartbeatCount() != 0 {
		t.Fatalf("unregistered heartbeat must not reach business listener")
	}
	if rogueConn.bytesWritten() != 0 {
		t.Fatalf("unregistered connection must not receive a heartbeat response")
	}
}

// 已注册 A 冒用 B 的 runnerId 发心跳：A、B 心跳时间都不变，且都不回包
func TestRegisteredConnectionCannotHeartbeatAsAnotherRunner(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	connA := newMockConn()
	ctxA := registerConn(t, dispatcher, sessions, connA, "A")
	connB := newMockConn()
	ctxB := registerConn(t, dispatcher, sessions, connB, "B")

	sessionA, _ := sessions.Get("A")
	sessionB, _ := sessions.Get("B")
	beforeA := sessionA.lastHeartbeatTime.Load()
	beforeB := sessionB.lastHeartbeatTime.Load()

	writtenABefore := connA.bytesWritten()
	writtenBBefore := connB.bytesWritten()

	dispatcher.Dispatch(ctxA, heartbeatEnvelope("B"))

	if sessionB.lastHeartbeatTime.Load() != beforeB {
		t.Fatalf("forged heartbeat must not refresh B")
	}
	if sessionA.lastHeartbeatTime.Load() != beforeA {
		t.Fatalf("forged heartbeat must not refresh A either")
	}
	if listener.heartbeatCount() != 0 {
		t.Fatalf("forged heartbeat must not reach business listener")
	}
	if connA.bytesWritten() != writtenABefore {
		t.Fatalf("forged heartbeat must not be answered to A")
	}
	if connB.bytesWritten() != writtenBBefore {
		t.Fatalf("B must not receive a response for A's forgery")
	}
	if ctxB.Session != sessionB {
		t.Fatalf("precondition: B context should hold session B")
	}
}

// 正常心跳：刷新自身会话并只回发给发送方
func TestNormalHeartbeatRefreshesOwnSession(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	conn := newMockConn()
	ctx := registerConn(t, dispatcher, sessions, conn, "B")
	sessionB, _ := sessions.Get("B")
	ageHeartbeat(sessionB, 5*time.Second)
	before := sessionB.lastHeartbeatTime.Load()
	writtenBefore := conn.bytesWritten()

	dispatcher.Dispatch(ctx, heartbeatEnvelope("B"))

	if sessionB.lastHeartbeatTime.Load() <= before {
		t.Fatalf("normal heartbeat should refresh its own session")
	}
	if listener.heartbeatCount() != 1 {
		t.Fatalf("normal heartbeat should reach business listener exactly once")
	}
	if conn.bytesWritten() <= writtenBefore {
		t.Fatalf("sender should receive the heartbeat response")
	}
	if sessionB.timedOut.Load() {
		t.Fatalf("session should not be marked timed out")
	}
}

// ---- 主动断开身份校验 ----

// 未注册连接冒用 B 身份发 DISCONNECT：B 会话不受影响
func TestUnregisteredConnectionCannotDisconnectRunner(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	registerConn(t, dispatcher, sessions, newMockConn(), "B")
	sessionB, _ := sessions.Get("B")
	listener.reset()

	rogueConn := newMockConn()
	rogueCtx := &ConnContext{Session: NewRunnerSession("", rogueConn, "", "", "")}
	dispatcher.Dispatch(rogueCtx, disconnectEnvelope("B", "bye"))

	if current, ok := sessions.Get("B"); !ok || current != sessionB {
		t.Fatalf("B session must survive forged disconnect")
	}
	if listener.disconnectCount("B") != 0 {
		t.Fatalf("business listener must not see B disconnect")
	}
	if !rogueConn.isClosed() {
		t.Fatalf("unregistered connection should be closed")
	}
}

// 已注册 A 冒用 B 身份发 DISCONNECT：只影响 A 自己，绝不移除 B
func TestRegisteredConnectionCannotDisconnectAnotherRunner(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	ctxA := registerConn(t, dispatcher, sessions, newMockConn(), "A")
	connA := ctxA.Session.Conn.(*mockConn)
	registerConn(t, dispatcher, sessions, newMockConn(), "B")
	sessionB, _ := sessions.Get("B")
	listener.reset()

	dispatcher.Dispatch(ctxA, disconnectEnvelope("B", "bye"))

	if current, ok := sessions.Get("B"); !ok || current != sessionB {
		t.Fatalf("forged disconnect must not remove B")
	}
	if listener.disconnectCount("B") != 0 {
		t.Fatalf("B must not be reported disconnected")
	}
	// 伪造身份的消息被拒绝：关闭发送方自己的连接（会话清理由连接退出流程负责）
	if !connA.isClosed() {
		t.Fatalf("forged disconnect should close the sender's own connection")
	}
}

// 正常主动断开：条件移除自身会话并通知一次
func TestNormalDisconnectRemovesOwnSessionAndNotifiesOnce(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	conn := newMockConn()
	ctx := registerConn(t, dispatcher, sessions, conn, "A")
	listener.reset()

	dispatcher.Dispatch(ctx, disconnectEnvelope("A", "shutdown"))

	if _, ok := sessions.Get("A"); ok {
		t.Fatalf("session should be removed on normal disconnect")
	}
	events := listener.disconnectsFor("A")
	if len(events) != 1 || events[0].reason != "shutdown" {
		t.Fatalf("expected exactly one shutdown disconnect event, got %+v", events)
	}
	if !conn.isClosed() {
		t.Fatalf("connection should be closed on normal disconnect")
	}
}

// ---- 回执与产物请求身份校验 ----

// 已注册 A 冒用 B 身份发状态/日志回执：业务层不以 B 身份处理
func TestRegisteredConnectionCannotForgeQueryResults(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	ctxA := registerConn(t, dispatcher, sessions, newMockConn(), "A")
	registerConn(t, dispatcher, sessions, newMockConn(), "B")
	listener.reset()

	dispatcher.Dispatch(ctxA, statusEnvelope("B"))
	dispatcher.Dispatch(ctxA, logsEnvelope("B"))

	if listener.statusCount() != 0 || listener.logCount() != 0 {
		t.Fatalf("forged query results must not be processed as B")
	}
}

// 已注册 A 用 B 的 source_id 请求产物：按连接绑定识别为 A，不以 B 身份回调
func TestRegisteredConnectionCannotForgeArtifactIdentity(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	ctxA := registerConn(t, dispatcher, sessions, newMockConn(), "A")
	registerConn(t, dispatcher, sessions, newMockConn(), "B")
	listener.reset()

	dispatcher.Dispatch(ctxA, artifactEnvelope("B", "svc-1"))

	if listener.artifactCount() != 0 {
		t.Fatalf("artifact request with mismatched source_id must be rejected")
	}
}

// 正常产物请求：按连接绑定识别身份
func TestNormalArtifactRequestIsAccepted(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	ctx := registerConn(t, dispatcher, sessions, newMockConn(), "A")
	listener.reset()

	dispatcher.Dispatch(ctx, artifactEnvelope("A", "svc-1"))

	if listener.artifactCount() != 1 {
		t.Fatalf("normal artifact request should reach business listener")
	}
}

// ---- 接管后的旧连接 ----

// 旧连接被同 ID 新连接接管后继续发送已注册消息：一律拒绝，新会话不受影响
func TestSupersededConnectionMessagesAreRejected(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	oldCtx := registerConn(t, dispatcher, sessions, newMockConn(), "R")
	stale := oldCtx.Session

	// 模拟新连接接管：注册表替换为新会话，旧 ConnContext 仍指向旧会话
	fresh := registeredSession(sessions, "R", newMockConn())
	ageHeartbeat(fresh, 5*time.Second)
	before := fresh.lastHeartbeatTime.Load()
	listener.reset()

	if sessions.IsCurrent(stale) {
		t.Fatalf("precondition: stale session should no longer be the current one")
	}

	dispatcher.Dispatch(oldCtx, heartbeatEnvelope("R"))
	dispatcher.Dispatch(oldCtx, disconnectEnvelope("R", "bye"))
	dispatcher.Dispatch(oldCtx, statusEnvelope("R"))
	dispatcher.Dispatch(oldCtx, artifactEnvelope("R", "svc-1"))

	if listener.heartbeatCount() != 0 {
		t.Fatalf("superseded connection heartbeat must be rejected")
	}
	if listener.statusCount() != 0 {
		t.Fatalf("superseded connection status must be rejected")
	}
	if listener.artifactCount() != 0 {
		t.Fatalf("superseded connection artifact request must be rejected")
	}
	if listener.disconnectCount("R") != 0 {
		t.Fatalf("superseded connection must not disconnect R")
	}
	if fresh.lastHeartbeatTime.Load() != before {
		t.Fatalf("new session must not be refreshed by the old connection")
	}
	if current, ok := sessions.Get("R"); !ok || current != fresh {
		t.Fatalf("new session must stay registered")
	}
}

// 接管后新会话一切正常
func TestSupersedingSessionKeepsWorking(t *testing.T) {
	sessions := NewSessionManager()
	listener := newRecordingListener(goodToken)
	dispatcher := newDispatchedMaster(sessions, listener)

	registerConn(t, dispatcher, sessions, newMockConn(), "R")

	// 新连接注册（走真实注册路径）
	newConn := newMockConn()
	newCtx := registerConn(t, dispatcher, sessions, newConn, "R")
	listener.reset()

	current, _ := sessions.Get("R")
	if current != newCtx.Session {
		t.Fatalf("registry should hold the new session")
	}

	writtenBefore := newConn.bytesWritten()
	dispatcher.Dispatch(newCtx, heartbeatEnvelope("R"))

	if listener.heartbeatCount() != 1 {
		t.Fatalf("new session should keep heartbeating")
	}
	if newConn.bytesWritten() <= writtenBefore {
		t.Fatalf("new session should receive its heartbeat response")
	}
}
